[Ryan Chung Yi Sheng] iP

NekrozQliphort opened the pull request:
## Duke ~~Silver aka Ron Swanson~~
> The art of war has two important lessons. One is to use Duke to solve your procrastination problems. Two, is to never quote something that someone never said. - Sun Tzu :smile:

Duke is a day-to-day helper, which attempts to help with:
- [X] Tracking Tasks 
- [X] Backing Up Tasks
- [X] Accompany you because you need **friends**

All you need to know about Duke is:
- Text-based
- Easy to Learn
- Fast to Use

To set up just:
1. Download it from [here](https://github.com/NekrozQliphort/ip)!
2. Click and enter.
3. Literally just play around with it.

Did you know that you can use \``` to fence your code?
Allow me to demonstrate with `main`:
```java
public class Main {
    public static void main(String[] args) {
        System.out.println("I want my reward!");
    }
}
```

Here's a reward for reaching the [end](https://youtu.be/dQw4w9WgXcQ)!

Commit:
Add Gradle support

NekrozQliphort committed:
Add greet, echo and exit functionality

NekrozQliphort committed:
Add the ability to add and list task, remove echo functionality

NekrozQliphort committed:
Add the task class with ability to mark as done

NekrozQliphort committed:
Remove redundant string format

NekrozQliphort committed:
Add task of type Todo

NekrozQliphort committed:
Add task of type Deadline

NekrozQliphort committed:
Add task of type Event

NekrozQliphort committed:
Add functionality to add todo, deadline and event type tasks

NekrozQliphort committed:
Change tabs into 4 spaces for output

NekrozQliphort committed:
Add ACTUAL.TXT to .gitignore

NekrozQliphort committed:
Add automated text UI testing

NekrozQliphort committed:
Add the functionality of handling errors and more inputs

NekrozQliphort committed:
Update ui-test

NekrozQliphort committed:
Add delete functionality with extra checks

NekrozQliphort committed:
Move printing functionality from Task to Duke and minor formatting

NekrozQliphort committed:
Add more tests for new functionalities

NekrozQliphort committed:
Modify modifier order to comply with style guide

NekrozQliphort committed:
Fix regex and add trailing spaces

NekrozQliphort committed:
Add TaskList with basic functionalities

NekrozQliphort committed:
Add backupFormat to format object to be stored

NekrozQliphort committed:
Add storage functionality

NekrozQliphort committed:
Update .gitignore to include data folder

NekrozQliphort committed:
Include storage functionality in Duke

NekrozQliphort committed:
Clean up storage constructor

NekrozQliphort committed:
Add generic Parser and DateTimeParser to parse strings

NekrozQliphort committed:
Add constuctors to accomodate datetime

NekrozQliphort committed:
Change Duke's functionalities to accomodate datetime

NekrozQliphort committed:
Add datetime functionality

NekrozQliphort committed:
Add datetime backup functionality

NekrozQliphort committed:
Add Ui and CommandParser

NekrozQliphort committed:
Fix CommandParser bugs

NekrozQliphort committed:
Modify test cases

NekrozQliphort committed:
Add duke package

NekrozQliphort committed:
Add tests for CommandParser

NekrozQliphort committed:
Fix bug for CommandParser

NekrozQliphort committed:
Create MANIFEST.MF

NekrozQliphort committed:
Fix duke so that it handles unknown commands properly

NekrozQliphort committed:
Reformat code to comply with coding standard

NekrozQliphort committed:
Add documentation for public classes and methods

NekrozQliphort committed:
Add find task functionality to Duke

NekrozQliphort committed:
Merge branch 'branch-Level-9'

NekrozQliphort committed:
Merge branch 'branch-A-JavaDoc'

NekrozQliphort committed:
Merge branch 'branch-A-CodingStandard'

simonjulianl commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+	private final static String[] COMMANDS_WITH_ARGS = new String[] {"todo", "deadline", 
+			"event", "delete", "done", "find"};
+	private final static String[] COMMANDS_WITHOUT_ARGS = new String[] {"list", "bye"};
+	private final static String UNKNOWN_COMMAND_MSG = "☹ OOPS!!! " +
```
Should we remove the plus to a new line, it's part of the coding standard

NekrozQliphort replied:
> Should we remove the plus to a new line, it's part of the coding standard

Yes it should. Do allow me to change it.

simonjulianl commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		for (String cmdCheck : COMMANDS_WITH_ARGS) {
+			isInvalidCommand = isInvalidCommand && !(cmdSplit[0].equals(cmdCheck));
```
Should we assign a variable to define a meaning of `!(cmdSplit[0].equals(cmdCheck))` ?

NekrozQliphort replied:
I do somewhat agree with this assessment. I have assigned `cmdSplit[0]` to a variable, but as for `cmdSplitFront.equals(cmdCheck)` as a whole, I believe it is only used in 2 places, so it might seem redundant. However, do let me know if you disagree.

simonjulianl replied:
I see, in my opinion its more like the same level of abstraction, if the left side is `isInvalidCommand`, the right side usually is not supposed to do some lower-level operations, it just makes it easier to read, not really about redundancy

simonjulianl commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+		}
+
+		switch (cmdSplit[0]) {
+		case "todo":
```
Shall we explicitly mention `Fallthrough` here, another suggestion is that you can combine this into a list and check if command in list

NekrozQliphort replied:
Yes, do allow me to change that.

simonjulianl commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,41 @@
+			try {
+				return LocalDate.parse(s, formatter);
+			} catch (DateTimeParseException e) {
+				// Don't need to do anything here
```
Shall we rename this to /* no-op */ ?

NekrozQliphort replied:
Good idea!

simonjulianl commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,77 @@
+ */
+public class Deadline extends Task {
+    private static final DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy");
+    protected String by;
```
Shall we make it private to keep it simple for now as no classes currently extends `Deadline`

NekrozQliphort replied:
Agreed. I shall change it.

simonjulianl commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,77 @@
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
+        this.byDate = null;
```
`this` is not required as its not shadowed

NekrozQliphort replied:
I added `this` for clarity as there is another constructor that has `byDate` as a parameter, and in my opinion, better indicates this is a object variable.

simonjulianl replied:
I see, but I think its just part of our coding guideline that if it's not shadowed, then we should not use null (or use null if and only if its shadowed) unfortunately

simonjulianl replied:
I see, I might have missed it

simonjulianl commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * A class representing a parser that returns a date.
+ */
+public class DateTimeParser extends Parser<LocalDate> {
```
I like how you have your own `Parser` with generic, I think it makes the code neat

simonjulianl commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,18 @@
+ * 
+ * @param <T> The type of Object for the string to be parsed into.
+ */
+public abstract class Parser<T> {
```
This is elegant, but shall we use `interface` instead of `abstract class` here as you are only defining behaviours here?

NekrozQliphort replied:
May I ask for the reason? In my opinion, in the case I want to extend it further to include some variables, which might be the case, it seems an abstract class would work better, but do let me know if you disagree.

simonjulianl replied:
hem its fine if you want to extend it in the future, I think its just my preference to not add something that I wont be certain I will add it or not (like just focus on the present first), but it really depends on the scope of the project

simonjulianl commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
Wah I really just know you can do this!

simonjulianl reviewed (changes requested):
Overall really clean and nice code! clear separation of concerns between classes!

okyntary commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+	private static final String MISSING_ARGUMENT_TEMPLATE = "☹ OOPS!!! " +
+			"Seems like there are missing argument(s) for %s";
+	private static final String NON_NUMBER_ARGUMENT_TEMPLATE = "☹ OOPS!!! " +
+			"Should have entered a number for %s, BOI";
```
Should the "+"s in lines 10, 12 and 14 be moved to be part of the following line, as part of the coding standard?

NekrozQliphort replied:
Yes, I believe so. Please allow me to change it.

okyntary commented on src/main/java/duke/CommandParser.java:
```diff
@@ -0,0 +1,74 @@
+		case "todo":
+		case "done":
+		case "delete":
+		case "find":
```
As Simon also said, the coding standard indicates that ```case``` without a ```break``` should have ```// Fallthough``` commented

NekrozQliphort replied:
Agreed, I shall change that.

okyntary commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,68 @@
+    protected String description;
+    protected boolean isDone;
+
+    /**a
```
Small typo here.

NekrozQliphort replied:
Thanks! Would not have caught that myself.

okyntary reviewed (commented):
Looks good to me. I don't have much to add about any coding standard violations.

NekrozQliphort committed:
Address issues mentioned by peers

NekrozQliphort committed:
Merge branch 'branch-PR-Issues'

NekrozQliphort committed:
Merge branch 'add-gradle-support'

NekrozQliphort committed:
Add checkstyle

NekrozQliphort committed:
Fix style according to checkstyle

NekrozQliphort committed:
Merge branch 'branch-A-CheckStyle'

NekrozQliphort committed:
Implement GUI for Duke

NekrozQliphort committed:
Merge branch 'branch-Level-10'

NekrozQliphort committed:
Fix checkstyle issues

NekrozQliphort committed:
Add assertions to CommandParser and MainWindow

NekrozQliphort committed:
CommandParser: Add cmdFormatValidate method as a helper method for parse

The method parse is long and the formatting can be abstracted into
another method.

Let's abstract the logic for formatting and validation into another
method.

This allows part of the method parse to be abstracted resulting in
shorter code.

NekrozQliphort committed:
Storage: change the default case for load to return empty list instead.

The expected behaviour for load is to return an empty list should the
formatting have any errors, as opposed to return a list with some tasks.

Changing the default case to return an empty list handles the case as
mentioned above.

Let's change the default case for load to return an empty list.

NekrozQliphort committed:
Remove unused code and add explanation comments

NekrozQliphort committed:
Merge pull request #1 from NekrozQliphort/branch-A-CodeQuality

Merge branch-A-CodeQuality into master

NekrozQliphort committed:
Merge branch 'master' into branch-A-Assertions

NekrozQliphort committed:
Merge pull request #2 from NekrozQliphort/branch-A-Assertions

Merge branch-A-Assertions into master

NekrozQliphort committed:
Fix bugs regarding list and bye commands not working

NekrozQliphort committed:
Implement checks to detect if tasks clash

NekrozQliphort committed:
Merge branch 'branch-B-DetectAnomalies'

NekrozQliphort committed:
Create a better GUI for Duke

NekrozQliphort committed:
Merge branch 'branch-A-BetterGui'

NekrozQliphort committed:
Change error message for non valid arguments for delete

NekrozQliphort committed:
Add more javadocs

NekrozQliphort committed:
Add screenshot and documentation

NekrozQliphort committed:
Reformat README.md

NekrozQliphort committed:
Change mainClassName in build.gradle to the correct class

NekrozQliphort committed:
Rename constants to comply with naming convention

NekrozQliphort committed:
Add index checks to handle out of bounds error for task list

NekrozQliphort committed:
Set title for Duke

NekrozQliphort committed:
Fix taskList index checking bug

NekrozQliphort committed:
Change image to avoid copyright
